pkg/provision/workspace/persistentHome.go

```diff
@@ -0,0 +1,87 @@
+//
+// Copyright (c) 2019-2023 Red Hat, Inc.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package workspace
```

Comment:
Minor nit: this function is more of a library function than a provision function -- we expect packages in `pkg/provision` to sync resources with the cluster, whereas `pkg/library` packages operate on DevWorkspaces themselves.

Reply:
I'd also consider not using dwerrors here out of principle, as it suggests a `retryError` may be returned which is not the case (we're just reusing it for warnings -- instead I'd return something like `(warning string, err error)`) but I'm not super firm on this one so it's up to you.

Structurally, _maybe_ also consider taking a "return a modified copy" over editing the object passed in -- we do both but IMO it's cleaner to return a copy that you have apply as we do in the flatten package.

Reply:
> Structurally, _maybe_ also consider taking a "return a modified copy" over editing the object passed in -- we do both but IMO it's cleaner to return a copy that you have apply as we do in the flatten package.

I reworked things to go with your suggested approach.

One thing I went back-and-fourth on while working on this was trying to hide the complexity of the different cases that can occur from the devworkspace controller code. With the latest revision, the controller code is relatively simple, but the return values of `AddPersistentHomeVolume()` are relatively complex (`bool, DWTemplateSpec and error`).

`AddPersistentHomeVolume()` could be simplified by making the controller call `needsPersistentHomeDirectory()` and deal with the different cases itself, but this made the controller code look a bit messy IMO. I'm open to going with this approach, however.